src/nl/hannahsten/texifyidea/util/files/Files.kt

```diff
@@ -106,9 +114,109 @@ fun Document.psiFile(project: Project): PsiFile? = PsiDocumentManager.getInstanc
  * @return The created file.
  */
 fun createFile(fileName: String, contents: String): File {
+    val currentFileName = getUniqueFileName(fileName)
+
+    return File(currentFileName).apply {
+        createNewFile()
+        LocalFileSystem.getInstance().refresh(true)
+        writeText(contents, StandardCharsets.UTF_8)
+    }
+}
+
+/**
+ * This writes to a file without using java.io.File
+ *
+ * Needs to be wrapped in a writable environment
+ *
+ * @param project The project this is for
+ * @param filePath The absolute path of the file to create. File extension provided later
+ * @param text The text to add to the file
+ * @param root The root path of the project? Could possibly be replaced
+ * @param extension The desired file extension **without** a dot, tex by default
+ *
+ * @return Returns the name of the file that was created without folder names, and with the extension
+ */
+fun writeToFileUndoable(project: Project, filePath: String, text: String, root: String, extension: String = "tex"): String {
+    val filenameNoExtension = Path.of(filePath).fileName.toString()
+    val filepathNoFilename = Path.of(filePath).parent.pathString
+
+    // Create file...but not on fs yet
+    val fileFactory = PsiFileFactory.getInstance(project)
+    val newfile = fileFactory.createFileFromText(
+        getUniqueFileName(
+            filenameNoExtension.appendExtension(extension),
+            filepathNoFilename
+        ),
+        LatexFileType,
+        text
+    )
+
+    val projectRootManager = ProjectRootManager.getInstance(project)
+    val allRoots = projectRootManager.contentRoots + projectRootManager.contentSourceRoots
+
+    // The following is going to resolve the PsiDirectory that we need to add the new file to.
+    var relativePath = ""
+    var bestRoot: VirtualFile? = null
+    for (testFile in allRoots) {
```

Comment:
Why do you check all content roots? As I understood, you just want to find the directory of the root file (say main.tex) right, since you are going to create subdirectories from there? So why don't you use `psiFile.findRootFile().parent`?

Reply:
Because it was adapted from [here](https://github.com/JetBrains/intellij-community/blob/0a0cb763e024b92140e8cbdd7b2340a1f37169dc/python/python-psi-impl/src/com/jetbrains/python/refactoring/classes/extractSuperclass/PyExtractSuperclassHelper.java#L201-L263) which did the same. I think the idea is we want to be flexible and just find the requested dir by running up from root. Doing so from the opened file is not guaranteed to be any better so I didn't make the adaptation since it works as is.

Reply:
Interesting, well I don't know how but it seems to work so should be fine :D